ftdc/ftdc_test.go

```diff
@@ -107,3 +108,133 @@ func TestFTDCSchemaGenerations(t *testing.T) {
 	test.That(t, datum2.Data["foo1"], test.ShouldResemble, map[string]float32{"X": 1, "Y": 0})
 	test.That(t, datum2.Data["foo2"], test.ShouldResemble, map[string]float32{"X": 2, "Y": 0})
 }
+
+type badStatser struct{}
+
+func (badStatser badStatser) Stats() any {
+	// Returning maps are disallowed.
+	return map[string]float32{"X": 42}
+}
+
+func TestRemoveBadStatser(t *testing.T) {
+	logger := logging.NewTestLogger(t)
+
+	// ftdcData will be appended to on each call to `writeDatum`. At the end of the test we can pass
+	// this to `parse` to assert we have the expected results.
+	ftdcData := bytes.NewBuffer(nil)
+	ftdc := NewWithWriter(ftdcData, logger.Sublogger("ftdc"))
+
+	// `foo` implements `Statser`.
+	foo1 := &foo{x: 1, y: 2}
+	ftdc.Add("foo1", foo1)
+
+	// `badStatser` implements `Statser`, but returns a map instead of a struct. This will fail at
+	// `writeDatum`.
+	ftdc.Add("badStatser", badStatser{})
+
+	// constructDatum should succeed as it does not perform validation.
+	datum := ftdc.constructDatum()
+	test.That(t, len(datum.Data), test.ShouldEqual, 2)
+	test.That(t, datum.Data["foo1"], test.ShouldNotBeNil)
+	test.That(t, datum.Data["badStatser"], test.ShouldNotBeNil)
+
+	// writeDatum will discover the map and error out.
+	err := ftdc.writeDatum(datum)
+	test.That(t, err, test.ShouldNotBeNil)
```

Comment:
If you were to call this again, is the expectation that it would not error and that it would self-heal? If so, can we test that?

Reply:
No. The datum is invalid. Think of it as immutable and permanently bad because it contains a map.

But the side-effect of `writeDatum` is to "heal" the set of statsers to use. So further calls to `constructDatum` create legit `datum`s.

Which is what this test is showing off. On line 151 after inspecting that this `err`ored out for the reason we expect.